File: jskat-swing-gui/src/main/java/org/jskat/gui/swing/iss/ISSTablePanel.java
package org.jskat.gui.swing.iss;

import com.google.common.eventbus.Subscribe;
import org.jskat.control.event.iss.IssTableGameStartedEvent;
import org.jskat.control.event.iss.IssTableStateChangedEvent;
import org.jskat.control.event.skatgame.GameStartEvent;
import org.jskat.control.gui.action.JSkatAction;
import org.jskat.data.iss.ChatMessage;
import org.jskat.data.iss.GameStartInformation;
import org.jskat.data.iss.PlayerStatus;
import org.jskat.data.iss.TablePanelStatus;
import org.jskat.gui.swing.LayoutFactory;
import org.jskat.gui.swing.table.ContextPanelType;
import org.jskat.gui.swing.table.JSkatUserPanel;
import org.jskat.gui.swing.table.OpponentPanel;
import org.jskat.gui.swing.table.SkatTablePanel;
import org.jskat.util.GameVariant;
import org.jskat.util.Player;

import javax.swing.*;
import java.util.Arrays;
import java.util.List;

/**
 * Panel for ISS table
 */
public class ISSTablePanel extends SkatTablePanel {

    ChatPanel chatPanel;

    // FIXME (jansch 05.04.2011) Dirty hack
    TablePanelStatus lastTableStatus;

    /**
     * Constructor
     *
     * @param tableName Table name
     * @param actions   Action map
     */
    public ISSTablePanel(final String tableName, final ActionMap actions) {
        super(tableName, actions);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected List<JSkatAction> getGameOverActions() {
        return Arrays.asList(
                JSkatAction.INVITE_ISS_PLAYER,
                JSkatAction.READY_TO_PLAY,
                JSkatAction.TALK_ENABLED,
                JSkatAction.LEAVE_ISS_TABLE);
    }

    /**
     * @see SkatTablePanel#getPlayGroundPanel()
     */
    @Override
    protected JPanel getPlayGroundPanel() {
        final JPanel panel = new JPanel(LayoutFactory.getMigLayout("fill,insets 0", "[grow][shrink]", "fill"));
        panel.add(super.getPlayGroundPanel(), "width 80%, grow");

        // replace game start context panel
        addContextPanel(ContextPanelType.START, new StartContextPanel(getActionMap(), getGameOverActions()));
        setContextPanel(ContextPanelType.START);

        return panel;
    }

    @Override
    protected JTabbedPane getLeftPanel() {
        JTabbedPane leftPanel = super.getLeftPanel();

        this.chatPanel = getChatPanel();
        this.chatPanel.addNewChat(
                this.strings.getString("table") + " " + getName(), getName());
        leftPanel.add(this.strings.getString("chat"), this.chatPanel);

        return leftPanel;
    }

    @Override
    protected OpponentPanel getOpponentPanel() {
        return new OpponentPanel(getActionMap(), 12, true);
    }

    @Override
    protected JSkatUserPanel createPlayerPanel() {
        return new JSkatUserPanel(getActionMap(), 12, true);
    }

    private ChatPanel getChatPanel() {
        return new ChatPanel(this);
    }

    @Override
    protected JPanel getRightPanelForTrickPanel() {
        final JPanel additionalActionsPanel = new JPanel(LayoutFactory.getMigLayout());
        additionalActionsPanel.setOpaque(false);

        final JButton resignButton = new JButton(getActionMap().get(JSkatAction.RESIGN));
        additionalActionsPanel.add(resignButton, "growx, wrap");
        final JButton showCardsButton = new JButton(getActionMap().get(JSkatAction.SHOW_CARDS));
        additionalActionsPanel.add(showCardsButton, "growx");

        return additionalActionsPanel;
    }

    /**
     * {@inheritDoc}
     */
    @Subscribe
    public void clearTableOn(IssTableGameStartedEvent event) {

        if (event.gameStart.loginName().equals(event.gameStart.playerNames().get(Player.FOREHAND))) {
            clearTable(event.tableName, Player.MIDDLEHAND, Player.REARHAND, Player.FOREHAND, event.gameStart);
        } else if (event.gameStart.loginName().equals(event.gameStart.playerNames().get(Player.MIDDLEHAND))) {
            clearTable(event.tableName, Player.REARHAND, Player.FOREHAND, Player.MIDDLEHAND, event.gameStart);
        } else if (event.gameStart.loginName().equals(event.gameStart.playerNames().get(Player.REARHAND))) {
            clearTable(event.tableName, Player.FOREHAND, Player.MIDDLEHAND, Player.REARHAND, event.gameStart);
        }
    }

    private void clearTable(String tableName, final Player leftOpponent, final Player rightOpponent,
                            final Player player, final GameStartInformation gameStart) {

        // FIXME: should have been sent via the event bus, event is missing in history
        resetTableOn(new GameStartEvent(gameStart.gameNo(), GameVariant.STANDARD, leftOpponent, rightOpponent, player));

        setPlayerName(leftOpponent, gameStart.playerNames().get(leftOpponent));
        setPlayerTime(leftOpponent, gameStart.playerTimes().get(leftOpponent));
        setPlayerName(rightOpponent, gameStart.playerNames().get(rightOpponent));
        setPlayerTime(rightOpponent, gameStart.playerTimes().get(rightOpponent));
        setPlayerName(player, gameStart.playerNames().get(player));
        setPlayerTime(player, gameStart.playerTimes().get(player));
    }

    /**
     * Updates the panel with the new status
     *
     * @param event Table status changed event
     */
    @Subscribe
    public void updateTableStatusOn(IssTableStateChangedEvent event) {

        TablePanelStatus tableStatus = event.status;

        // FIXME (jansch 05.04.2011) make 3<>4 change possible
        // setMaxPlayers(tableStatus.getMaxPlayers());

        for (final String playerName : tableStatus.getPlayerInformation().keySet()) {

            final PlayerStatus status = tableStatus.getPlayerInformation(playerName);

            if (!status.isPlayerLeft()) {
                addPlayerName(playerName);
            }
            setPlayerReadyToPlay(playerName, status.isReadyToPlay());
            setPlayerChatEnabled(playerName, status.isTalkEnabled());
            if (status.isPlayerLeft()) {
                removePlayerName(playerName);
            }
        }

        this.lastTableStatus = tableStatus;
    }

    private void addPlayerName(final String playerName) {

        if (!this.playerNamesAndPositions.containsKey(playerName)) {

            this.playerNamesAndPositions.put(playerName, null);

            if (this.userPanel.getPlayerName() == null) {

                this.userPanel.setPlayerName(playerName);

            } else if (this.leftOpponentPanel.getPlayerName() == null) {

                this.leftOpponentPanel.setPlayerName(playerName);

            } else if (this.rightOpponentPanel.getPlayerName() == null) {

                this.rightOpponentPanel.setPlayerName(playerName);
            }
        }
    }

    private void removePlayerName(final String playerName) {

        if (this.playerNamesAndPositions.containsKey(playerName)) {

            this.playerNamesAndPositions.remove(playerName);

            if (playerName.equals(this.userPanel.getPlayerName())) {

                this.userPanel.setPlayerName("");

            } else if (playerName.equals(this.leftOpponentPanel.getPlayerName())) {

                this.leftOpponentPanel.setPlayerName("");

            } else if (playerName.equals(this.rightOpponentPanel.getPlayerName())) {

                this.rightOpponentPanel.setPlayerName("");
            }
        }
    }

    /**
     * Adds a new chat message to the chat
     *
     * @param message Chat message
     */
    public void appendChatMessage(final ChatMessage message) {
        this.chatPanel.appendMessage(message);
    }
}

File: jskat-gui/src/main/kotlin/org/jskat/gui/javafx/table/ISSTablePanel.kt
package org.jskat.gui.javafx.table

import com.google.common.eventbus.Subscribe
import javafx.application.Platform
import javafx.scene.Node
import javafx.scene.control.Button
import javafx.scene.layout.HBox
import javafx.scene.layout.Priority
import javafx.scene.layout.VBox
import org.jskat.control.event.iss.IssTableGameStartedEvent
import org.jskat.control.event.iss.IssTableStateChangedEvent
import org.jskat.control.event.skatgame.GameStartEvent
import org.jskat.control.gui.action.JSkatAction
import org.jskat.control.gui.action.JSkatActionEvent
import org.jskat.data.iss.ChatMessage
import org.jskat.data.iss.GameStartInformation
import org.jskat.data.iss.TablePanelStatus
import org.jskat.gui.action.AbstractJSkatAction
import org.jskat.gui.javafx.iss.ChatPanel
import org.jskat.gui.javafx.iss.IssStartContextPanel
import org.jskat.util.GameVariant
import org.jskat.util.Player

class ISSTablePanel(tableName: String, actions: Map<JSkatAction, AbstractJSkatAction>) : SkatTablePanel(tableName, actions) {

    private lateinit var chatPanel: ChatPanel
    private var lastTableStatus: TablePanelStatus? = null

    override fun initPanel() {
        // Don't call super.initPanel() as we are replacing the entire layout

        chatPanel = ChatPanel(actions)
        chatPanel.addNewChat(strings.getString("table") + " " + tableName, tableName)

        val playGroundPanel = super.getPlayGroundPanel()
        val issContent = HBox(playGroundPanel, chatPanel)
        HBox.setHgrow(playGroundPanel, Priority.ALWAYS)

        center = issContent

        // Replace the default start panel with the ISS version
        val issStartActions = listOf(
            JSkatAction.INVITE_ISS_PLAYER,
            JSkatAction.READY_TO_PLAY,
            JSkatAction.TALK_ENABLED,
            JSkatAction.LEAVE_ISS_TABLE
        )
        val issStartPanel = IssStartContextPanel(actions, issStartActions)
        addContextPanel(ContextPanelType.START, issStartPanel)
        setContextPanel(ContextPanelType.START)
    }

    override fun getOpponentPanel(): OpponentPanel {
        return OpponentPanel(12, true)
    }

    override fun createPlayerPanel(): JSkatUserPanel {
        return JSkatUserPanel(12, true, actions)
    }

    override fun getRightPanelForTrickPanel(): Node {
        val additionalActionsPanel = VBox()
        additionalActionsPanel.spacing = 10.0

        val resignAction = actions[JSkatAction.RESIGN]
        if (resignAction != null) {
            val resignButton = Button(resignAction.getValue(AbstractJSkatAction.NAME) as? String ?: JSkatAction.RESIGN.name)
            resignButton.setOnAction {
                resignAction.actionPerformed(JSkatActionEvent(JSkatAction.RESIGN, it.source))
            }
            additionalActionsPanel.children.add(resignButton)
        }

        val showCardsAction = actions[JSkatAction.SHOW_CARDS]
        if (showCardsAction != null) {
            val showCardsButton = Button(showCardsAction.getValue(AbstractJSkatAction.NAME) as? String ?: JSkatAction.SHOW_CARDS.name)
            showCardsButton.setOnAction {
                showCardsAction.actionPerformed(JSkatActionEvent(JSkatAction.SHOW_CARDS, it.source))
            }
            additionalActionsPanel.children.add(showCardsButton)
        }

        return additionalActionsPanel
    }

    @Subscribe
    fun clearTableOn(event: IssTableGameStartedEvent) {
        val gameStart = event.gameStart
        val userLogin = gameStart.loginName
        val playerNames = gameStart.playerNames

        var userPosition: Player? = null
        for ((key, value) in playerNames) {
            if (value == userLogin) {
                userPosition = key
                break
            }
        }

        if (userPosition != null) {
            val leftOpponent = userPosition.leftNeighbor
            val rightOpponent = userPosition.rightNeighbor
            clearTable(leftOpponent, rightOpponent, userPosition, gameStart)
        }
    }

    private fun clearTable(
        leftOpponent: Player,
        rightOpponent: Player,
        player: Player,
        gameStart: GameStartInformation
    ) {
        resetTableOn(GameStartEvent(gameStart.gameNo, GameVariant.STANDARD, leftOpponent, rightOpponent, player))

        setPlayerName(leftOpponent, gameStart.playerNames[leftOpponent]!!)
        setPlayerTime(leftOpponent, gameStart.playerTimes[leftOpponent]!!)

        setPlayerName(rightOpponent, gameStart.playerNames[rightOpponent]!!)
        setPlayerTime(rightOpponent, gameStart.playerTimes[rightOpponent]!!)

        setPlayerName(player, gameStart.playerNames[player]!!)
        setPlayerTime(player, gameStart.playerTimes[player]!!)
    }

    @Subscribe
    fun updateTableStatusOn(event: IssTableStateChangedEvent) {
        val tableStatus = event.status

        Platform.runLater {
            for ((playerName, status) in tableStatus.playerInformation) {
                if (!status.isPlayerLeft) {
                    addPlayerName(playerName)
                }
                setPlayerReadyToPlay(playerName, status.isReadyToPlay)
                setPlayerChatEnabled(playerName, status.isTalkEnabled)
                if (status.isPlayerLeft) {
                    removePlayerName(playerName)
                }
            }
            lastTableStatus = tableStatus
        }
    }

    private fun addPlayerName(playerName: String) {
        if (!playerNamesAndPositions.containsKey(playerName)) {
            if (userPanel.playerName == null || userPanel.playerName!!.isEmpty()) {
                userPanel.playerName = playerName
                userPanel.position?.let { playerNamesAndPositions[playerName] = it }
            } else if (leftOpponentPanel.playerName == null || leftOpponentPanel.playerName!!.isEmpty()) {
                leftOpponentPanel.playerName = playerName
                leftOpponentPanel.position?.let { playerNamesAndPositions[playerName] = it }
            } else if (rightOpponentPanel.playerName == null || rightOpponentPanel.playerName!!.isEmpty()) {
                rightOpponentPanel.playerName = playerName
                rightOpponentPanel.position?.let { playerNamesAndPositions[playerName] = it }
            } else {
                playerNamesAndPositions[playerName] = null
            }
        }
    }

    private fun removePlayerName(playerName: String) {
        if (playerNamesAndPositions.containsKey(playerName)) {
            playerNamesAndPositions.remove(playerName)
            when (playerName) {
                userPanel.playerName -> userPanel.playerName = ""
                leftOpponentPanel.playerName -> leftOpponentPanel.playerName = ""
                rightOpponentPanel.playerName -> rightOpponentPanel.playerName = ""
            }
        }
    }

    private fun setPlayerName(player: Player, playerName: String) {
        Platform.runLater {
            super.getHandPanel(player).playerName = playerName
        }
    }

    private fun setPlayerReadyToPlay(playerName: String, readyToPlay: Boolean) {
        // Platform.runLater {
            super.getHandPanel(playerName)?.setReadyToPlay(readyToPlay)
        // }
    }

    private fun setPlayerChatEnabled(playerName: String, chatEnabled: Boolean) {
        // Platform.runLater {
            super.getHandPanel(playerName)?.setChatEnabled(chatEnabled)
        // }
    }

    fun appendChatMessage(message: ChatMessage) {
        chatPanel.appendMessage(message)
    }

    fun getChatPanel(): ChatPanel {
        return chatPanel
    }
}
